controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -243,41 +272,82 @@ private Job deploymentJob(Application application, JobType jobType, String reaso
     private List<Job> computeReadyJobs(ApplicationId id) {
         List<Job> jobs = new ArrayList<>();
         applications().get(id).ifPresent(application -> {
-            List<DeploymentSpec.Step> steps = application.deploymentSpec().steps().isEmpty()
-                    ? Collections.singletonList(new DeploymentSpec.DeclaredZone(Environment.test))
+            List<Step> steps = application.deploymentSpec().steps().isEmpty()
+                    ? singletonList(new DeploymentSpec.DeclaredZone(test))
                     : application.deploymentSpec().steps();
+            List<Step> productionSteps = steps.stream().filter(step -> step.deploysTo(prod) || step.zones().isEmpty()).collect(toList());
 
-            Optional<Instant> completedAt = Optional.of(clock.instant());
-            String reason = "Deploying " + application.change();
+            Optional<Instant> completedAt = Optional.of(Instant.EPOCH); // Delays before first production job are ignored.
+            String reason = "New change available";
+            List<Job> testJobs = null;
 
-            for (DeploymentSpec.Step step : steps) {
+            for (Step step : productionSteps) {
                 Set<JobType> stepJobs = step.zones().stream().map(order::toJob).collect(toSet());
-                Set<JobType> remainingJobs = stepJobs.stream().filter(job -> ! completedAt(application.change(), application, job).isPresent()).collect(toSet());
-                if (remainingJobs.isEmpty()) { // All jobs are complete -- find the time of completion of this step.
+                Map<Optional<Instant>, List<JobType>> jobsByCompletion = stepJobs.stream().collect(groupingBy(job -> completedAt(application.change(), application, job)));
```

Comment:
Optional keys is rarely a good idea. If it's optional the key shouldn't have been added to the map in the first place. If you need to do something for the empty value, use the "zero" value of the type instead (e.g. `Instant.EPOCH` in this case).

Reply:
Rarely, but in this case I think it's OK: the map is used only as a local variable, so it doesn't leak out anywhere; and the key is exactly an Optional<Instant> -- it represents both whether its value is complete, and if it is, when. 

I prefer this over generating the values several times, since I prefer to compute things _once_, and reuse the results. (Efficiency is one consideration, but perhaps more important is that you do things right both or none of the times.)